src/transformation/visitors/language-extensions/operators.ts

```diff
@@ -0,0 +1,176 @@
+    const signature = context.checker.getResolvedSignature(node);
+    if (!signature || !signature.declaration) {
+        return;
+    }
+    const typeDeclaration = getTypeDeclaration(signature.declaration);
+    if (!typeDeclaration) {
+        return;
+    }
+    const mapping = extensions.getExtensionKind(typeDeclaration);
+    if (mapping !== undefined && operatorMapExtensions.has(mapping)) {
+        return mapping;
+    }
+}
+
+function isOperatorMapDeclaration(declaration: ts.Declaration) {
+    const typeDeclaration = getTypeDeclaration(declaration);
+    if (typeDeclaration) {
+        const extensionKind = extensions.getExtensionKind(typeDeclaration);
+        return extensionKind !== undefined ? operatorMapExtensions.has(extensionKind) : false;
+    }
+}
+
+function isOperatorMapType(context: TransformationContext, type: ts.Type): boolean {
+    if (type.isUnionOrIntersection()) {
+        return type.types.some(t => isOperatorMapType(context, t));
+    } else {
+        return type.symbol?.declarations?.some(isOperatorMapDeclaration);
+    }
+}
+
+function isOperatorMapIdentifier(context: TransformationContext, node: ts.Identifier) {
+    const type = context.checker.getTypeAtLocation(node);
+    return isOperatorMapType(context, type);
+}
+
+export function isOperatorMapping(context: TransformationContext, node: ts.CallExpression | ts.Identifier) {
+    if (ts.isCallExpression(node)) {
+        return getOperatorMapExtensionKindForCall(context, node) !== undefined;
+    } else {
+        return isOperatorMapIdentifier(context, node);
+    }
+}
+
+export function transformOperatorMappingExpression(
+    context: TransformationContext,
+    node: ts.CallExpression
+): lua.Expression {
+    const extensionKind = getOperatorMapExtensionKindForCall(context, node);
+    assert(extensionKind);
+
+    const isBefore53 =
+        context.luaTarget === LuaTarget.Lua51 ||
+        context.luaTarget === LuaTarget.Lua52 ||
+        context.luaTarget === LuaTarget.LuaJIT ||
+        context.luaTarget === LuaTarget.Universal;
+    if (isBefore53) {
+        const luaTarget = context.luaTarget === LuaTarget.Universal ? LuaTarget.Lua51 : context.luaTarget;
+        if (bitwiseOperatorMapExtensions.has(extensionKind)) {
+            context.diagnostics.push(unsupportedForTarget(node, "Native bitwise operations", luaTarget));
+        } else if (
+            extensionKind === extensions.ExtensionKind.FloorDivisionOperatorType ||
+            extensionKind === extensions.ExtensionKind.FloorDivisionOperatorMethodType
+        ) {
+            context.diagnostics.push(unsupportedForTarget(node, "Floor division operator", luaTarget));
+        }
+    }
+
+    const args = node.arguments.slice();
+    if (binaryOperatorMappings.has(extensionKind)) {
+        if (args.length === 1) {
+            assert(ts.isPropertyAccessExpression(node.expression) || ts.isElementAccessExpression(node.expression));
```

Comment:
Asserts should probably be removed. If there is still a chance the condition can occur it should be handled properly.

Reply:
The only way these would happen is if someone was messing with the extension declarations themselves. But, since TS won't be happy without them, I'll add in some proper diagnostics.

Reply:
I removed the assertions, but didn't add diagnostics since I can't even create tests for them. That would require somehow modifying the extension declarations for the tests. I think it's safe to let things just crash once users start messing with TSTL internals.